### DISCUSSION TITLE
Solis hybrid multiple charge/discharge time

### DISCUSSION THREAD

**Discussion by mart-mac** [wills106/homeassistant-solax-modbus#350] [General]
On the solis hybrid, you can set 3 different charge / discharge time windows.  I've searched the net for the different modbus address, but can't find them.  Anyone got this setup to be able to set the 3 different times?

**Comment by TartanEnergy** [comment]
Think I saw them in the solis plugin - custom_components/solax_modbus/plugin_solis.py

**Comment by alienatedsec** [comment]
I think @wills106 could extend the registers and include more in the integration.

It currently stops on discharge time `43150` as per the below, but it goes beyond that
https://github.com/wills106/homeassistant-solax-modbus/blob/37dc385e7404a4397bb3241e2beeb03e81ed28dc/custom_components/solax_modbus/plugin_solis.py#L1625-L1633

```
DEBUG:pysolarmanv5.pysolarmanv5:SENT: a5 17 00 10 45 00 00 32 c3 fc f3 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 01 03 a8 85 00 0a f4 44 c5 15
DEBUG:pysolarmanv5.pysolarmanv5:RECD: a5 27 00 10 15 00 be 32 c3 fc f3 02 01 0d cd 62 00 91 21 00 00 86 ed aa 63 01 03 14 01 f4 00 00 00 00 00 1e 00 04 00 1e 00 00 00 00 00 00 00 00 a0 1d 69 15
[500, 0, 0, 30, 4, 30, 0, 0, 0, 0]
```
The above output of my inverter
`500` - general/global charge current regardless of time - `43141`
`0` - general/global discharge current regardless of time - `43142`
`0` - 1# charge hour start time - `43143`
`30` - 1# charge min start time - `43144`
`4` - 1# charge hour end time - `43145`
`30` - 1# charge min end time - `43146`
`0` - 1# discharge hour start time - `43147`
`0` - 1# discharge min start time - `43148`
`0` - 1# discharge hour end time - `43149`
`0` - 1# discharge min end time - `43150`

```
 DEBUG:pysolarmanv5.pysolarmanv5:SENT: a5 17 00 10 45 00 00 32 c3 fc f3 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 01 03 a8 8f 00 0a d4 46 b1 15
DEBUG:pysolarmanv5.pysolarmanv5:RECD: a5 27 00 10 15 00 c7 32 c3 fc f3 02 01 30 cd 62 00 b4 21 00 00 86 ed aa 63 01 03 14 00 00 00 00 00 00 00 05 00 00 00 06 00 00 00 00 00 00 00 00 84 c2 17 15
[0, 0, 0, 5, 0, 6, 0, 0, 0, 0]
```
`0` - Not sure why - `43151`
`0` - Not sure why - `43152`
`0` - 2# charge hour start time - `43153`
`5` - 2# charge min start time - `43154`
`0` - 2# charge hour end time - `43155`
`6` - 2# charge min end time - `43156`
`0` - 2# discharge hour start time - `43157`
`0` - 2# discharge min start time - `43158`
`0` - 2# discharge hour end time - `43159`
`0` - 2# discharge min end time - `43160`

```
DEBUG:pysolarmanv5.pysolarmanv5:SENT: a5 17 00 10 45 00 00 32 c3 fc f3 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 01 03 a8 99 00 0a 35 82 58 15
DEBUG:pysolarmanv5.pysolarmanv5:RECD: a5 27 00 10 15 00 e1 32 c3 fc f3 02 01 92 cd 62 00 17 22 00 00 86 ed aa 63 01 03 14 00 00 00 00 00 00 00 07 00 00 00 08 00 00 00 08 00 00 00 0a ae 26 9b 15
[0, 0, 0, 7, 0, 8, 0, 8, 0, 10]
```
`0` - Not sure why - `43161`
`0` - Not sure why - `43162`
`0` - 3# charge hour start time - `43163`
`7` - 3# charge min start time - `43164`
`0` - 3# charge hour end time - `43165`
`8` - 3# charge min end time - `43166`
`0` - 3# discharge hour start time - `43167`
`8` - 3# discharge min start time - `43168`
`0` - 3# discharge hour end time - `43169`
`10` - 3# discharge min end time - `43170`

@wills106 I am happy to submit a pull request after testing or let you include the above holding registers - let me know

**Reply by wills106** [reply]
I was going to add them in, but I think at the time there wasn't many people with the Solis Inverter. So I stopped until more feedback.

Feel free to submit a PR.

**Reply by alienatedsec** [reply]
Done some initial testing an found some issues :) as always - I will edit (amend) the above register ranges

**Reply by mart-mac** [reply]
It would be nice to have them added, I'm sure for you knowledgeable people it would be 10mins work, probably take me hours

**Reply by alienatedsec** [reply]
Updated my comment and register ranges to reflect the below screen readings - I will raise PR @wills106 sometime this afternoon - will be going out with the family.
![image](https://user-images.githubusercontent.com/73167064/224542478-ebb9b45d-ae94-4634-b21d-da9973074149.png)

**Reply by alienatedsec** [reply]
@wills106 I am not sure why and if my approach is even good, but I declared something wrong or misunderstood the way it works. However, it fails at line 609 of the `plugin_solis.py`.

[plugin_solis.zip](https://github.com/wills106/homeassistant-solax-modbus/files/10951975/plugin_solis.zip)

I will check again tomorrow or you can give me a hint where I messed it up.

**Reply by wills106** [reply]
On first look at line 394 add a new line with
),

So line 395 = ),

**Reply by wills106** [reply]
You also need to add the equivalent of:

```
SolisModbusButtonEntityDescription( 
        name = "Update Charge/Discharge Times",
        key = "update_charge_discharge_times",
        register = 43143,
        allowedtypes = HYBRID,
        write_method = WRITE_MULTI_MODBUS,
        icon = "mdi:battery-clock",
        value_function = value_function_timingmode,
    ),
```

For each new time slot group.

If ```def value_function_timingmode``` doesn't work for all three groups?

**Reply by alienatedsec** [reply]
> So line 395 = ),

Found this - you are right, I just missed that. Thanks

> You also need to add the equivalent of:
> 
> ```
> SolisModbusButtonEntityDescription( 
>         name = "Update Charge/Discharge Times",
>         key = "update_charge_discharge_times",
>         register = 43143,
>         allowedtypes = HYBRID,
>         write_method = WRITE_MULTI_MODBUS,
>         icon = "mdi:battery-clock",
>         value_function = value_function_timingmode,
>     ),
> ```
> 
> For each new time slot group.

I did, I think. I updated the function to include additional keys (lines  92-117) and after adding 395 - updated registers in lines 1880 - 2042

I will test this shortly

**Reply by alienatedsec** [reply]
I need to amend those as well (92-117) when double-checked

**Reply by wills106** [reply]
As long as they are in a sequence they should be fine in a single "SolisModbusButtonEntityDescription" I haven't checked the doc's in a while so I can't verify.

Good luck with the testing.

**Reply by alienatedsec** [reply]
@wills106 So, I have done some little work and testing - four things I noticed

1.  We have this issue where the registers are not pulled from the inverter - I suggest having a button that could pull and provide current figures for timing slots. We discussed it somewhere already.

2.  I created additional timeslots and two additional buttons to update those timing slots - two reasons for this approach - if any of those timeslots are not defined (default 0) the timings updates will fail if only one timeslot was defined. The error is below.
```
2023-03-13 09:02:39.912 ERROR (MainThread) [custom_components.solax_modbus] cannot treat payload scale None SolisModbusNumberEntityDescription(key='timed_charge_start_h_2', device_class=None, entity_category=<EntityCategory.CONFIG: 'config'>, entity_registry_enabled_default=True, entity_registry_visible_default=True, force_update=False, icon='mdi:battery-clock', has_entity_name=False, name='Timed Charge Start Hours 2', translation_key=None, unit_of_measurement=None, max_value=None, min_value=None, native_max_value=23, native_min_value=0, native_unit_of_measurement=<UnitOfTime.HOURS: 'h'>, native_step=1, step=None, allowedtypes=4096, register=None, read_scale_exceptions=None, fmt='i', scale=1, state=None, max_exceptions=None, min_exceptions_minus=None, blacklist=None, write_method=3, initvalue=0, unit='_uint16')
2023-03-13 09:05:37.950 ERROR (MainThread) [custom_components.solax_modbus] cannot treat payload scale None SolisModbusNumberEntityDescription(key='timed_charge_start_h_3', device_class=None, entity_category=<EntityCategory.CONFIG: 'config'>, entity_registry_enabled_default=True, entity_registry_visible_default=True, force_update=False, icon='mdi:battery-clock', has_entity_name=False, name='Timed Charge Start Hours 3', translation_key=None, unit_of_measurement=None, max_value=None, min_value=None, native_max_value=23, native_min_value=0, native_unit_of_measurement=<UnitOfTime.HOURS: 'h'>, native_step=1, step=None, allowedtypes=4096, register=None, read_scale_exceptions=None, fmt='i', scale=1, state=None, max_exceptions=None, min_exceptions_minus=None, blacklist=None, write_method=3, initvalue=0, unit='_uint16')
```
I had to set those values to whatever number and could revert to 0 as it was initially, and then updating timings started working again. If we had a button to read those - we could define and avoid the above error, and potentially introduce only one button to update timings.

> As long as they are in a sequence they should be fine in a single "SolisModbusButtonEntityDescription"

It looks like there is a gap in registers, so there is no sequence. Registers `43151`, `43152`, `43161`, and `43162` equal `0`, and are not defined in the same way as `43141` and `43142`, which are the global current values for charge and discharge.

3.  I had an instance where timings could not be defined on the inverter - it failed when tried to save them. I took the picture before submitting `save`. It looks like if times of charge and discharge are overlapping (two bottom lines of the screenshot - timeslot 3 in the integration) then, it fails to save the settings. I am not sure how the inverter will react when trying to save overlapping values via registers. There would need to be some logic around to avoid overlapping and that is a long journey (at least for me)

![image](https://user-images.githubusercontent.com/73167064/224669992-11461ec2-ab3c-4508-80ce-89eb0c8603d7.png)

Have a look at the code here:
https://github.com/alienatedsec/homeassistant-solax-modbus/compare/main...solis_timeslots

4.  I don't know why, but I don't like it, and would not recommend integrating without logic, and better formatting behind it. I am not sure how you see it, but there are a lot of `data` to digest even though I believe I have it organised already with sections etc.

![image](https://user-images.githubusercontent.com/73167064/224678119-2fe426f1-4676-4ea9-808c-9d6d0c59ebe4.png)

**Reply by alienatedsec** [reply]
According to the below screenshot from the Solis app, there are more charge/discharge currents (timeslots 2 and 3) that could be defined, and those cannot be set on the inverter.

![image](https://user-images.githubusercontent.com/73167064/225605882-8482b25d-536b-4290-a3f5-c4429a0606e5.png)

I might include them in the fork and see how they work.

**Reply by mart-mac** [reply]
They don't do anything on the inverter, you can change them on the app, but nothing updates on the inverter

**Reply by alienatedsec** [reply]
Those are still there, so I guess will be considered for those timelines.

**Reply by mart-mac** [reply]
Do you mean they are there on the inverter, hidden away?

**Reply by mart-mac** [reply]
I've just added this multiple Charge and discharge times, but I'm getting an error on the "solis Update Charge/Discharge Times 2" & 3 press button. Any ideas?

**Failed to call service button/press. int () argument must be a string, a bytes-like object or a real number, not 'NoneType'**

**Reply by alienatedsec** [reply]
@mart-mac change each value to some number and back to what you wish those to be set on. Those are not defined at first and need to be initially set for the sync button to work.

**Reply by mart-mac** [reply]
Thanks,  that was an easy fix

**Comment by wills106** [comment]
1. I had to decouple the sensors from the number entities as the Integration would replace your new values before you had chance to program them. Ideally I need to find a way to only update the sensors if the values actually change and not on each poll cycle.
2. Not really sure if one button for all is best or one per time slot?
3. I have noticed on other Inverters they don't like overlap. I'm not really sure if logic should go into the Integration. I think that belongs in Automations. The Integration should just provide the connection.
4. I agree it's all a mess. Unfortunately at the moment there isn't the equivalent of input_datetime at the moment, hopefully that gets added into HA/Core soon.

**Reply by wills106** [reply]
Just noticed you have changed block_size from 48 to 40
Did you test it at 48 first? As I made changes in 2023.03.2b3 that might allow the block_size to stay at 48 and handle bad reads in another way (ignore_readerror = True,)

**Reply by alienatedsec** [reply]
> Did you test it at 48 first?

I haven't tested at 48 blocks as 40 blocks worked fine for me with an earlier version - I was to wait for #341 to resolve first.
